### PR TITLE
The project structure was refactored

### PR DESCRIPTION
**Pull request by Fat-Frumos**

### CODE REVIEW

**Comment by sslichnyi** [comment]
@SpringBootApplication(scanBasePackages = {"com.epam.esm"}). scanBasePackages is redundant

**Comment by sslichnyi** [comment]
### data generation:
- it is not good idea to take data from remove repository for testing of fulfilling db with mock data. it is needed than it is better to move those files to this project (this repo).
- IOFile can be replaces with Files.readLines(). Please read about Java NIO. https://attacomsian.com/blog/java-read-write-files-nio-api


* Please, remove it. 


* it is better to create a new project which will be responsible of generation of mock data. I recommend to create a new project and generate data into csv file, into data.sql file (with insert sql queries), or in json file. Then make DataLoader to insert data into DB from this file. 


* It is not in scope of this MJC program. Data generation is not main task to do. But if you are interested in how to generate data it can be discussed and reviewed, but it must be done in separate project or module.

**Comment by sslichnyi** [comment]
### RequestErrorAttributes
- What is the purpose of RequestErrorAttributes? Why it is implemented? What do this do?

**Comment by sslichnyi** [comment]
### ErrorHandlerController:


- handleCertificateIsExistsException() and handleTagIsExistsException() methods can be merged into single method like this:


```
@ExceptionHandler({ CertificateAlreadyExistsException.class, TagAlreadyExistsException.class})
public Error... badRequest() {
....
}
```

**Comment by sslichnyi** [comment]
### EnumConverterFactory:

* There is no need to convert String to some enum class. Jackson do this by itself. It is like out-of-the-box-feature of Jackson.
Please, try to do that: create dto with an enum class as fields and take a look if it works or not. 

* If it is already implemented that leave it as it is.

* P.S. it is very cool that you are able to implement such functionality. if one of requirement will be to parse String value to enum in case insensitive mode than you will be able to do it without any doubts, but right now there is no such requirements so it is time cost.

**Comment by sslichnyi** [comment]
### RestController annotation:

- There is no need to specify `produces = APPLICATION_JSON_VALUE` because it is default value when RestController annotation is specified.

- Remove it everywhere.

- Please, take a look on this search: https://www.google.com/search?q=What+is+the+default+media+type+in+Spring+RestController%3F&rlz=1C5CHFA_enUA976UA976&sxsrf=APwXEddyzIZmlU3gqpl5HQbyr2-zylR3cA%3A1684501653520&ei=lXRnZKW1H5G43APn_JnoBA&ved=0ahUKEwjl07nVuYH_AhURHHcKHWd-Bk0Q4dUDCA8&uact=5&oq=What+is+the+default+media+type+in+Spring+RestController%3F&gs_lcp=Cgxnd3Mtd2l6LXNlcnAQAzIHCCEQoAEQCjIHCCEQoAEQCjIHCCEQoAEQCjIHCCEQoAEQCjoKCAAQRxDWBBCwA0oECEEYAFD6A1j6A2DYBmgBcAF4AIABmQGIAZkBkgEDMC4xmAEAoAECoAEByAEIwAEB&sclient=gws-wiz-serp

**Comment by sslichnyi** [comment]
### Controllers

- `public EntityModel<Linkable> getCertificateById(...) {...}` - it is not clear what return type is. It doesn't clear which dto java object will be return.
- Please, replace Linkable interface with the concrete java object everywhere.

**Comment by sslichnyi** [comment]
### Pageable

- In `public CollectionModel<EntityModel<Linkable>> getAll(...) {...}` method `SortOrder sort, FilterParams params, int size, int page` parameters are specified. It is already implemented. All this parameters must be replaces into single `Pageable pageable` parameter.  it is very simular of what you already have `Criteria` object. 
- Take a look on this examples:
    -  https://reflectoring.io/spring-boot-paging/

**Comment by sslichnyi** [comment]
`@GetMapping("")` is can be simplified to `@GetMapping`

**Comment by sslichnyi** [comment]
After import `tomcat.postman_collection.json` I was not able to use POST for certificates. Please take a look on it

**Comment by sslichnyi** [comment]
###  Clean Code

- `mapper.toDto(certificateDao.getById(id)
.orElseThrow(() -> new CertificateNotFoundException(
String.format("%s id: %d", MESSAGE, id))))` - please, split it into several lines, it is not clear what is returned from dao and what class is passed to mapper.
- Also split it in other places

**Comment by sslichnyi** [comment]
### DRY principle

- `update(
  @RequestBody final CertificateDto dto,
  @PathVariable final Long id)` - id field is specified two time (in dto and in path variables). it is confusing which one will be taken.
- Please, use only one approach

**Comment by sslichnyi** [comment]
### Performance Issues

- two calls to db is done: 1 - get entity, 2 - delete entity. it is performance issue because entity may contains a lot of information and it will take a lot of time to process it. 
- Actually you have to do two steps: 1 - map to Entity object, 2 - delete it by entityManager. it will be enough. 
- You also call getById() method in other places which is not correct.
- Please, fix it.

**Comment by sslichnyi** [comment]
### Single Responsibility Principle in CertificateDaoImpl method

- `public void delete(final Long id)` - this method do two things (`entityManager.find()` and `entityManager.remove()`) but it must to do only one of it.
- Please, refactor it.

**Comment by sslichnyi** [comment]
Note for me:
- requirements where not ensured all
- API end points where not tested

**Comment by Fat-Frumos** [comment]
> ### Single Responsibility Principle in CertificateDaoImpl method
> * `public void delete(final Long id)` - this method do two things (`entityManager.find()` and `entityManager.remove()`) but it must to do only one of it.
> * Please, refactor it.

> Removed .find(id) from Dao Layer in delete() method

**Comment by Fat-Frumos** [comment]
> ### DRY principle
> * `update( @RequestBody final CertificateDto dto, @PathVariable final Long id)` - id field is specified two time (in dto and in path variables). it is confusing which one will be taken.
> * Please, use only one approach

> Removed `@PathVariable final Long id`
> Changed the signature method to `update(Certificate certificate)`

**Comment by Fat-Frumos** [comment]
> ### Clean Code
> * `mapper.toDto(certificateDao.getById(id) .orElseThrow(() -> new CertificateNotFoundException( String.format("%s id: %d", MESSAGE, id))))` - please, split it into several lines, it is not clear what is returned from dao and what class is passed to mapper.
> * Also split it in other places

> Separating a nested entity in a new line

**Comment by Fat-Frumos** [comment]
> ### Performance Issues
> * two calls to db is done: 1 - get entity, 2 - delete entity. it is performance issue because entity may contains a lot of information and it will take a lot of time to process it.
> * Actually you have to do two steps: 1 - map to Entity object, 2 - delete it by entityManager. it will be enough.
> * You also call getById() method in other places which is not correct.
> * Please, fix it.

>   Getting a reference to an entity without loading its data with `entityManager.getReference(Entity.class, id);`.

**Comment by Fat-Frumos** [comment]
> ### RequestErrorAttributes
> * What is the purpose of RequestErrorAttributes? Why it is implemented? What do this do?

> RequestErrorAttributes is redundant, this class has been removed

> ### EnumConverterFactory:
> * There is no need to convert String to some enum class. Jackson do this by itself. It is like out-of-the-box-feature of Jackson.
>   Please, try to do that: create dto with an enum class as fields and take a look if it works or not.
> * If it is already implemented that leave it as it is.
> * P.S. it is very cool that you are able to implement such functionality. if one of requirement will be to parse String value to enum in case insensitive mode than you will be able to do it without any doubts, but right now there is no such requirements so it is time cost.

> Removed EnumConverterFactory

**Comment by Fat-Frumos** [comment]
> ### ErrorHandlerController:
> * handleCertificateIsExistsException() and handleTagIsExistsException() methods can be merged into single method like this:
> 
> ```
> @ExceptionHandler({ CertificateAlreadyExistsException.class, TagAlreadyExistsException.class})
> public Error... badRequest() {
> ....
> }
> ```

> AlreadyExistsExceptions have been joined into one method
> NotFoundExceptions have also joined

**Comment by Fat-Frumos** [comment]
> ### Controllers
> * `public EntityModel<Linkable> getCertificateById(...) {...}` - it is not clear what return type is. It doesn't clear which dto java object will be return.
> * Please, replace Linkable interface with the concrete java object everywhere.

> Interface Linkable has been removed
> RepresentationModelAssembler converts a specific dto object

> simplified `@GetMapping`

>removed `@SpringBootApplication(scanBasePackages = {"com.epam.esm"})`

**Comment by Fat-Frumos** [comment]
> ### Pageable
> * In `public CollectionModel<EntityModel<Linkable>> getAll(...) {...}` method `SortOrder sort, FilterParams params, int size, int page` parameters are specified. It is already implemented. All this parameters must be replaces into single `Pageable pageable` parameter.  it is very simular of what you already have `Criteria` object.
> * Take a look on this examples:
>   
>   * https://reflectoring.io/spring-boot-paging/

> The Criteria class to the Pageable interface has been changed.
> The `getAll(Pageable pageable)` method in the Service layer returns the Page interface instead of the List

**Comment by Fat-Frumos** [comment]
> ### RestController annotation:
> * There is no need to specify `produces = APPLICATION_JSON_VALUE` because it is the default value when RestController annotation is specified.
> * Remove it everywhere.

> removed all unnecessary `produces = APPLICATION_JSON_VALUE` in Controllers class

**Comment by Fat-Frumos** [comment]
> ### Data generation:
> * it is not good idea to take data from remove repository for testing of fulfilling db with mock data. it is needed than it is better to move those files to this project (this repo).
> * Please, remove it.
> * IOFile can be replaces with Files.readLines(). > * > * it is better to create a new project which will be responsible of generation of mock data. I recommend to create a new project and generate data into csv file, into data.sql file (with insert sql queries), or in json file. Then make DataLoader to insert data into DB from this file.
> * It is not in scope of this MJC program. Data generation is not main task to do. But if you are interested in how to generate data it can be discussed and reviewed, but it must be done in separate project or module.

> Removed classes DataLoader, EntityBuilder
> Added SQL files, used `public DataSourceInitializer dataSourceScriptInitializer` to set data
> Updated Id generator strategy (gift_certificate_id_seq)
> `@GeneratedValue(strategy = SEQUENCE, generator = "certificate_sequence")
>  @SequenceGenerator(name = "certificate_sequence", sequenceName = "gift_certificate_id_seq", allocationSize = 1)`

**Comment by Fat-Frumos** [comment]
> After import `tomcat.postman_collection.json` I was not able to use POST for certificates. 

> Updated `tomcat.postman_collection.json`: 
> - created js-tests for the collection, 
> - delimiter on get, added tree-like folders

**Comment by Fat-Frumos** [comment]
Added new tests
The problem of multiple retrievals from the database is solved
Updated collections, added tests

File: Module#03/web-app/src/main/java/com/epam/esm/config/PersistenceConfig.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.Optional;
+import java.util.Properties;
+
+/**
+ * Configuration class for persistence-related beans.
+ */
+@Configuration
+@RequiredArgsConstructor
+@EnableTransactionManagement
+public class PersistenceConfig {
```
Comment (sslichnyi) [inline Module#03/web-app/src/main/java/com/epam/esm/config/PersistenceConfig.java]: it is better to move this database config to repository impl module

File: Module#03/web-app/src/main/java/com/epam/esm/controller/CertificateController.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+ * {@link RestController} to indicate that it is a Spring MVC controller,
+ * {@link RequestMapping} with a value of "/certificates" to map requests,
+ * and {@link CrossOrigin} with origins set to "*"
+ * and allowed headers set to "GET", "POST", "PUT", and "DELETE"
+ * to enable Cross-Origin Resource Sharing (CORS).
+ */
+@RestController
+@RequiredArgsConstructor
+@RequestMapping(value = "/certificates")
+@CrossOrigin(origins = "*", allowedHeaders = {"GET", "POST", "PUT", "DELETE"})
```
Comment (sslichnyi) [inline Module#03/web-app/src/main/java/com/epam/esm/controller/CertificateController.java]: if I am not wrong, those CrossOrigin is configured in WebMvcConfig class. So maybe this annotation is not needed here

**Comment by sslichnyi** [comment]
It is also good to test REST application with Layered approach (Slices). With WebMockMvcTest, SpringBootTest and DataJpaTest annotations.

**Comment by sslichnyi** [comment]
Business requirements:
- [ ] Change single field of gift certificate (e.g. implement the possibility to change only duration of a certificate or only price).
- [x] Add new entity User.
- [x] implement only get operations for user entity.
- [x] Make an order on gift certificate for a user (user should have an ability to buy a certificate).
- [x] Get information about user’s orders.
- [x] Get information about user’s order: cost and timestamp of a purchase.
- [x] The order cost should not be changed if the price of the gift certificate is changed.
- [x] Get the most widely used tag of a user with the highest cost of all orders.
- [x] Create separate endpoint for this query.
- [x] Demonstrate SQL execution plan for this query (explain).
- [x] Search for gift certificates by several tags (“and” condition).
- [x] Pagination should be implemented for all GET endpoints. Please, create a flexible and non-erroneous solution. Handle all exceptional cases.
- [x] Support HATEOAS on REST endpoints.

Application requirements
- [x] JDK version: 8. Use Streams, java.time.*, an etc. where it is appropriate. (the JDK version can be increased in agreement with the mentor/group coordinator/run coordinator)
- [x] Application packages root: com.epam.esm.
- [x] Java Code Convention is mandatory (exception: margin size –120 characters).
- [x] Apache Maven/Gradle, latest version. Multi-module project.
- [x] Spring Framework, the latest version.
- [x] Database: PostgreSQL/MySQL, latest version.
- [x] Testing: JUnit, the latest version, Mockito.
- [ ] Service layer should be covered with unit tests not less than 80%. - coverage is 70%
- [x] Hibernate should be used as a JPA implementation for data access.
- [x] Spring Transaction should be used in all necessary areas of the application.
- [ ] Audit data should be populated using JPA features (an example can be found in materials).

**Comment by sslichnyi** [comment]
- Change single field of gift certificate (e.g. implement the possibility to change only duration of a certificate or only price).

Those two requirements were not implemented

**Review by sslichnyi — APPROVED** [review]

**Comment by Fat-Frumos** [comment]
> * Change single field of gift certificate (e.g. implement the possibility to change only duration of a certificate or only price).
> 
> Those two requirements were not implemented

> Added PatchDto with fields id, duration, price, tags

> Created AuditListener with preUpdate, prePersist methods

File: Module#03/web-app/src/main/java/com/epam/esm/controller/CertificateController.java (hunk 2 above)
Comment (Fat-Frumos) [reply Module#03/web-app/src/main/java/com/epam/esm/controller/CertificateController.java]: > if I am not wrong, those CrossOrigin is configured in WebMvcConfig class. So maybe this annotation is not needed here

> Duplicate annotation removed

**Comment by Fat-Frumos** [comment]
> It is also good to test REST application with Layered approach (Slices). With WebMockMvcTest, SpringBootTest and DataJpaTest annotations.

> Updated WebMvcTest for controllers, created CertificateApiTest Integration test

**Pull request merged**